[ENG-9064][1/2] EAS Update: a basic debugging guide, try 2 (#23079)
expo/expo: docs/pages/eas-update/debug.mdx

---
title: Debug EAS Update
sidebar_title: Basic
description: Learn how to debug EAS Update.
---

import ImageSpotlight from '~/components/plugins/ImageSpotlight';
import { Terminal } from '~/ui/components/Snippet';

> This guide shows how to verify our configuration so that we can find the source of problems like an app not showing a published update. It's important to tell the current state of our app at any given time, so we built EAS Update with this in mind. Once we know which updates are running on which builds, we can make changes so that our apps are in the state we expect.

Reviewer: Multiple personality 'we' in this sentence as well.

## Go to the Deployments page

> The EAS website has a [Deployments page](https://expo.dev/accounts/[account]/projects/[project]/deployments) that shows the current state of our app. The term _deployment_ refers to a group of builds and their updates. If we've made builds and updates with EAS, you can see your project's status on the website in the Deployments tab.

Reviewer 2: ```suggestion
The EAS website has a [Deployments page](https://expo.dev/accounts/[account]/projects/[project]/deployments) that shows the current state of our app. The term _deployment_ refers to a group of builds and their corresponding updates. If we've made builds and updates with EAS, we can see our project's status on the website in the Deployments tab.
```

(fwiw, I don't really like the `we` voice for docs in general, but that's outside the scope of this PR. this suggestion just adheres to a consistent voice)

<ImageSpotlight
  alt="Deployments tab"
  src="/static/images/eas-update/deployments-website-tab.png"
  style={{ maxWidth: 400 }}
/>

## Common problems

The following section describes common problems and how to fix them. Below is a diagram of how EAS Update works and the spots that are useful to inspect when finding the root cause of an issue. In the following sections, we'll inspect and verify these spots and more.

<ImageSpotlight alt="Map of debugging spots" src="/static/images/eas-update/debug-map.png" />

### Unexpected channel

<ImageSpotlight

>   alt="Deployment has no updates"

Reviewer: Fix various alt texts on the different screenshots throughout.

  src="/static/images/eas-update/deployments-wrong-channel.png"
  style={{ maxWidth: 600 }}
/>

> If the deployment channel is unexpected, it means your build was not built with the correct channel. To fix this, [configure your channel](/#configure-channel).

Reviewer 2: ```suggestion
If the deployment channel is unexpected, it means our build was not built with the correct channel. To fix this, [configure our channel](/#configure-channel).
```

> If the deployment channel is unexpected, it means your build was not built with the correct channel. To fix this, [configure your channel](/#configure-channel).

Reviewer: Emphasize that a rebuild of the app is necessary after changing this (common mistake).

### Unexpected runtime version

<ImageSpotlight
  alt="Deployment has no updates"
  src="/static/images/eas-update/deployments-wrong-runtime.png"
  style={{ maxWidth: 600 }}
/>

> If the deployment runtime version is unexpected, it means your build was not built with the correct runtime version. To fix this, [configure your runtime version](/#configure-runtime-version).

Reviewer 2: ```suggestion
If the deployment runtime version is unexpected, it means our build was not built with the correct runtime version. To fix this, [configure our runtime version](/#configure-runtime-version).
```

### Unexpected branch

<ImageSpotlight
  alt="Deployment has no updates"
  src="/static/images/eas-update/deployments-wrong-branch.png"
  style={{ maxWidth: 600 }}
/>

> If the deployment has an unexpected branch, you will need to [map your channel to the correct branch](/#map-channel-to-branch).

Reviewer 2: ```suggestion
If the deployment has an unexpected branch, we will need to [map our channel to the correct branch](/#map-channel-to-branch).
```

### Missing updates

<ImageSpotlight
  alt="Deployment has no updates"
  src="/static/images/eas-update/deployments-no-updates.png"
  style={{ maxWidth: 600 }}
/>

> The displayed deployment does not have any updates. To fix this, [publish an update to the branch](/#publish-update).

Reviewer 2: I like this voice much more than the `we` fwiw.

> The displayed deployment does not have any updates. To fix this, [publish an update to the branch](/#publish-update).

Reviewer: Other reason is that the runtime version _of the update_ doesn't match that of the build. This is common enough to note here as well.

### Missing branch

<ImageSpotlight
  alt="Deployment has no updates"
  src="/static/images/eas-update/deployments-no-branch.png"
  style={{ maxWidth: 600 }}
/>

The displayed deployment has the correct channel, but it is not linked to a branch. To fix this, [map the channel to the correct branch](/#map-channel-to-branch).

### Missing deployment

> If your deployment is not displayed, this means your build is not configured properly with EAS Update. To fix this, [configure your channel](/#configure-channel) and [configure your runtime version](/#configure-runtime-version).

Reviewer 2: ```suggestion
If your deployment is not displayed, this means our build is not configured properly with EAS Update. To fix this, [configure our channel](/#configure-channel) and [configure our runtime version](/#configure-runtime-version).
```

> If your deployment is not displayed, this means your build is not configured properly with EAS Update. To fix this, [configure your channel](/#configure-channel) and [configure your runtime version](/#configure-runtime-version).

Reviewer: ```suggestion
If your deployment is not displayed, this means your build is not configured properly for EAS Update. To fix this, [configure your channel](/#configure-channel) and [configure your runtime version](/#configure-runtime-version).
```

Maybe mention as well to verify the general EAS Update configuration (updates.url)?

## Solutions

### Configure channel

> To verify that a build has a specific channel, make sure that in **eas.json**, there is a channel property:

Reviewer: ```suggestion
To verify that a build has a specific channel, make sure that your build profile in **eas.json** has a channel property:
```

Then, we can run a command like `eas build --profile preview` to create a build with a channel named "preview".

### Configure runtime version

> To verify your runtime version, make sure your app config (**app.json**/**app.config.js**) has a `runtimeVersion` property:

Reviewer 2: ```suggestion
To verify your runtime version, make sure our app config (**app.json**/**app.config.js**) has a `runtimeVersion` property:
```

> To verify your runtime version, make sure your app config (**app.json**/**app.config.js**) has a `runtimeVersion` property:

Reviewer: ```suggestion
To verify our runtime version, we make sure our app config (**app.json**/**app.config.js**) has a `runtimeVersion` property:
```

```json app.json
{
  "expo": {
    "runtimeVersion": {
      "policy": "sdkVersion"
    }
  }
}
```

> By default, it is `{ "policy": "sdkVersion" }`, but you can change your runtime to [use a different policy or a specific version](/eas-update/runtime-versions).

Reviewer 2: ```suggestion
By default, it is `{ "policy": "sdkVersion" }`, but we can change our runtime to [use a different policy or a specific version](/eas-update/runtime-versions).
```

  ]}
/>

> If your branch is not listed, you can create a new branch with `eas branch:create`

Reviewer 2: ```suggestion
If your branch is not listed, we can create a new branch with `eas branch:create`
```

> If your branch is not listed, you can create a new branch with `eas branch:create`

Reviewer: ```suggestion
If our branch is not listed, we can create a new branch with `eas branch:create`
```

### Publish update

To create and publish an update, we can run the following command:

<Terminal cmd={['$ eas update']} />

After publishing, the output will display the branch and the runtime version. This information can help us verify that we're creating an update with the configuration we expect.

> ## Mitigation steps

Reviewer: This is now a bit disconnected from the previous section. Any 'root cause of the issue' we might have found, was a configuration one, not an issue with the actual update up to this point.

While this will prevent all new users from seeing the bad update, users who've already received the bad update will run it until they can download the latest update. Since mobile networks are not always able to download the most recent update, sometimes users may run a bad update for a long time. When viewing error logs for our app, it's normal to see a lingering long tail of errors as our users' apps get the most recent update or build. We'll know we solved the bug when we see the error rate decline dramatically; however, it likely will not disappear completely if we have a diverse user base across many locations and mobile networks.

## Next steps

> Read the [advanced guide](/eas-update/debug-advanced) to debug EAS Update when the basic steps don't work.

Reviewer: ```suggestion
Read the [advanced guide](/eas-update/debug-advanced) to debug EAS Update when the basic steps don't resolve your issue.
```
